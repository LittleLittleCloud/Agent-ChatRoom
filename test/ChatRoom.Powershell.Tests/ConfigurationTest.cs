using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using ApprovalTests.Namers;
using ApprovalTests.Reporters;
using ApprovalTests;
using Json.Schema;
using Xunit;
using Json.Schema.Generation;
using FluentAssertions;

namespace ChatRoom.Powershell.Tests;

public class ConfigurationTest
{
    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ApprovalTests")]
    public void VerifyConfigurationSchema()
    {
        var schema = new JsonSchemaBuilder()
            .FromType<PowershellConfiguration>()
            .Build();

        var json = JsonSerializer.Serialize(schema, new JsonSerializerOptions { WriteIndented = true });
        var schemaFileName = "chatroom_powershell_configuration_schema.json";
        var schemaFilePath = Path.Join("template", "chatroom.powershell", schemaFileName);
        var schemaFile = File.ReadAllText(schemaFilePath);

        Approvals.Verify(json);
        schemaFile.Should().BeEquivalentTo(json);

        var command = new CreateConfigurationCommand();
        var schemaContent = command.GetSchemaContent();
        schemaContent.Should().BeEquivalentTo(json);
    }


    [Fact]
    [UseReporter(typeof(DiffReporter))]
    [UseApprovalSubdirectory("ApprovalTests")]
    public void VerifyAvailableTemplates()
    {
        var command = new CreateConfigurationCommand();
        var availableTemplates = command.AvailableTemplates;
        availableTemplates.Should().BeEquivalentTo(["chatroom-powershell"]);

        var listTemplatesCommand = new ListTemplatesCommand();
        listTemplatesCommand.AvailableTemplates.Keys.Should().BeEquivalentTo(availableTemplates);

        var templates = new List<string>();
        foreach (var template in availableTemplates)
        {
            var templateContent = command.GetTemplateContent(template);
            templates.Add(templateContent);
        }

        Approvals.VerifyAll("templates", templates, "templates");
    }
}
