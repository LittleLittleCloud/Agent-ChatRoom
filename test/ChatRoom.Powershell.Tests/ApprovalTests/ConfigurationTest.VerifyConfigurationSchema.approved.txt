{
  "type": "object",
  "properties": {
    "runner": {
      "type": "object",
      "properties": {
        "name": {
          "$ref": "#/$defs/string"
        },
        "description": {
          "$ref": "#/$defs/string1"
        },
        "last_n_message": {
          "$ref": "#/$defs/integer"
        }
      },
      "description": "Powershell runner configuration"
    },
    "gpt": {
      "type": "object",
      "properties": {
        "name": {
          "$ref": "#/$defs/string2"
        },
        "CurrentWorkingDirectory": {
          "$ref": "#/$defs/string3"
        },
        "system_message": {
          "$ref": "#/$defs/string4"
        },
        "description": {
          "$ref": "#/$defs/string5"
        },
        "openai_config": {
          "$ref": "#/$defs/openAIClientConfiguration"
        }
      },
      "description": "Powershell GPT configuration"
    },
    "room_config": {
      "type": "object",
      "properties": {
        "room": {
          "$ref": "#/$defs/string14"
        },
        "port": {
          "$ref": "#/$defs/integer1"
        }
      },
      "description": "The configuration for the chat room"
    }
  },
  "$defs": {
    "string": {
      "type": "string",
      "description": "Name of the powershell runner agent, default is \u0027ps-runner\u0027"
    },
    "string1": {
      "type": "string",
      "description": "Agent description of the powershell runner agent, default is \u0027A powershell script runner\u0027"
    },
    "integer": {
      "type": "integer",
      "description": "Number of last message to look for powershell script, default is 10"
    },
    "string2": {
      "type": "string",
      "description": "Name of the powershell gpt agent, default is \u0027ps-gpt\u0027"
    },
    "string3": {
      "type": "string",
      "description": "Current working directory, default is current directory of where the program is running"
    },
    "string4": {
      "type": "string",
      "description": "System message, default is \u0027$\u0022\u0022\u0022\r\n            You are a powershell developer. You need to convert the task assigned to you to a powershell script.\r\n            \r\n            If there is bug in the script, you need to fix it.\r\n\r\n            The current working directory is {cwd}\r\n\r\n            You need to write powershell script to resolve task. Put the script between \u0060\u0060\u0060pwsh and \u0060\u0060\u0060.\r\n            The script should always write the result to the output stream using Write-Host command.\r\n\r\n            e.g.\r\n            \u0060\u0060\u0060pwsh\r\n            # This is a powershell script\r\n            Write-Host \u0022Hello, World!\u0022\r\n            \u0060\u0060\u0060\r\n            \u0022\u0022\u0022\u0027"
    },
    "string5": {
      "type": "string",
      "description": "Agent description, default is \u0027I am PowerShell GPT, I am good at writing powershell scripts.\u0027"
    },
    "openAIClientConfiguration": {
      "type": "object",
      "properties": {
        "llm_type": {
          "$ref": "#/$defs/lLMType"
        },
        "openai_api_key": {
          "$ref": "#/$defs/string6"
        },
        "openai_model_id": {
          "$ref": "#/$defs/string7"
        },
        "azure_openai_endpoint": {
          "$ref": "#/$defs/string8"
        },
        "azure_openai_key": {
          "$ref": "#/$defs/string9"
        },
        "azure_openai_deploy_name": {
          "$ref": "#/$defs/string10"
        },
        "third_party_llm_endpoint": {
          "$ref": "#/$defs/string11"
        },
        "third_party_llm_key": {
          "$ref": "#/$defs/string12"
        },
        "third_party_llm_model_id": {
          "$ref": "#/$defs/string13"
        }
      },
      "description": "OpenAI configuration"
    },
    "lLMType": {
      "enum": [
        "AOAI",
        "OpenAI",
        "ThirdParty"
      ],
      "description": "LLM provider, default is Azure OpenAI"
    },
    "string6": {
      "type": "string",
      "description": "OpenAI API key, default is $env:OPENAI_API_KEY"
    },
    "string7": {
      "type": "string",
      "description": "OpenAI model ID, default is gpt-3.5-turbo"
    },
    "string8": {
      "type": "string",
      "description": "Azure OpenAI endpoint, default is $env:AZURE_OPENAI_ENDPOINT"
    },
    "string9": {
      "type": "string",
      "description": "Azure OpenAI key, default is $env:AZURE_OPENAI_API_KEY"
    },
    "string10": {
      "type": "string",
      "description": "Azure OpenAI deploy name, default is $env:AZURE_OPENAI_DEPLOY_NAME"
    },
    "string11": {
      "type": "string",
      "description": "Third-party LLM provider endpoint, default is $env:THIRD_PARTY_LLM_ENDPOINT"
    },
    "string12": {
      "type": "string",
      "description": "Third-party LLM provider key, default is $env:THIRD_PARTY_LLM_KEY"
    },
    "string13": {
      "type": "string",
      "description": "Third-party LLM provider model ID, default is $env:THIRD_PARTY_LLM_MODEL_ID"
    },
    "string14": {
      "type": "string",
      "description": "The name of the room. Default is \u0027room\u0027"
    },
    "integer1": {
      "type": "integer",
      "description": "The port number where the room is hosted. Default is 30000"
    }
  }
}